### PR TITLE
Fix setting connect timeout

### PR DESCRIPTION
**Pull request by glowdan**
1. fixed the bug of timeout set
2. add more options(persistent)

hope it's useful.

### CODE REVIEW

File: lib/Connection.php [hunk 1]
```diff
@@ -30,6 +30,8 @@ class Connection {
     private $uri;
     /** @var int */
     private $timeout = 1000;
+    /** @var int */
+    private $persistent = 0;
```
Comment (kelunik) [inline lib/Connection.php]: Why is that a `int`, not a `bool`?

File: lib/Connection.php [hunk 2]
```diff
@@ -249,7 +256,7 @@ private function closeSocket() {
         $this->outputBuffer = "";
         $this->outputBufferLength = 0;
 
-        if (is_resource($this->socket)) {
+        if (!$this->persistent && is_resource($this->socket)) {
```
Comment (kelunik) [inline lib/Connection.php]: `closeSocket` should still close the connection. Any explicit close must close the connection.

File: lib/Connection.php (hunk 1 above)
Comment (glowdan) [inline lib/Connection.php]: At the begining, I wish to set the flags. You give me a good suggest, it should only check the persistent.

File: lib/Connection.php (hunk 2 above)
Comment (glowdan) [inline lib/Connection.php]: I have test it, if the socket is closed, the persistent connection will be use in one request. We need use the connection in the children fpm process.

Comment (bwoebi) [inline lib/Connection.php]: The `fclose()` should be executed on explicit `close()` call, but not when called via `__destruct()`, if persistent.

Comment (glowdan) [inline lib/Connection.php]: I don't want to close the connection, the connection will exist until the fpm process closed.

Comment (bwoebi) [inline lib/Connection.php]: Correct. That's why fclose() should not be executed in destructor, only when you are explicitly closing it via a close() call.

Comment (glowdan) [inline lib/Connection.php]: If I do not call the close function, the progress will run forever,  and it won't execute the code after the amp\run() function. So I must call the close. 
If I call the` fclose()`, the persistent connection will lost. 
So...

Comment (kelunik) [inline lib/Connection.php]: Just call `Amp\stop`.

Comment (kelunik) [inline lib/Connection.php]: Could you separate the timeout fix into its own commit and PR from the other changes?

Comment (glowdan) [inline lib/Connection.php]: Thank you, I solve it. The Amp\stop is what I need.

Comment (glowdan) [inline lib/Connection.php]: Wait for a minute.

**Comment by kelunik** [comment]
Thanks!
